mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,253 @@
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        let io = Pin::new(&mut self.1);
+        match io.poll_write(self.0, buf) {
+            Poll::Ready(Ok(n)) => Ok(n),
+            Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+            Poll::Pending => Err(Error::SslWantWrite),
+        }
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite + 'static> Context<T> {
+    pub async fn establish_async<IoType>(&mut self, io: T, hostname: Option<&str>) -> Result<()>
+    where
+        for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<IoType>,
+    {
+        self.prepare_handshake(io, hostname)?;
+
+        struct HandshakeFuture<'a, T>(&'a mut Context<T>);
+        impl<T> Future for HandshakeFuture<'_, T>
+        where
+            for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+        {
+            type Output = Result<()>;
+            fn poll(mut self: Pin<&mut Self>, ctx: &mut TaskContext) -> std::task::Poll<Self::Output> {
+                self.0
+                    .with_bio_async(ctx, |ssl_ctx| match ssl_ctx.handshake() {
+                        Err(Error::SslWantRead) | Err(Error::SslWantWrite) => Poll::Pending,
+                        Err(e) => Poll::Ready(Err(e)),
+                        Ok(()) => Poll::Ready(Ok(())),
+                    })
+                    .unwrap_or(Poll::Ready(Err(Error::NetSendFailed)))
+            }
+        }
+
+        HandshakeFuture(self).await
+    }
+}
+
+impl<T: AsyncRead> AsyncRead for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &mut ReadBuf<'_>) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        self.with_bio_async(cx, |ssl_ctx| match ssl_ctx.recv(buf.initialize_unfilled()) {
+            Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(())),
+            Err(Error::SslWantRead) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+            Ok(i) => {
+                buf.advance(i);
+                Poll::Ready(Ok(()))
+            }
+        })
+        .unwrap_or_else(|| Poll::Ready(Err(crate::private::error_to_io_error(Error::NetRecvFailed))))
+    }
+}
+
+impl<T: AsyncWrite + Unpin> AsyncWrite for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_write(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &[u8]) -> Poll<IoResult<usize>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        let result = self
+            .with_bio_async(cx, |ssl_ctx| {
+                ssl_ctx.write_tracker.adjust_buf(buf)?;
```

Comment:
This ``write_tracker`` call advances ``buf`` by returning it; why aren't you capturing that return here? Document with a comment if OK to discard.

Reply:
fixed